src/Build/BackEnd/Components/RequestBuilder/ItemBucket.cs

```diff
@@ -95,11 +98,20 @@ private ItemBucket()
             }
 
             _metadata = metadata;
-            _expander = new Expander<ProjectPropertyInstance, ProjectItemInstance>(_lookup, _lookup, new StringMetadataTable(metadata), FileSystems.Default);
+            _expander = new Expander<ProjectPropertyInstance, ProjectItemInstance>(_lookup, _lookup, new StringMetadataTable(metadata), FileSystems.Default, loggingContext);
 
             _bucketSequenceNumber = bucketSequenceNumber;
         }
 
+        /// <summary>
+        /// Updates the logging context that this bucket is going to use.
+        /// </summary>
+        /// <param name="loggingContext"></param>
+        internal void UpdateLoggingContext(LoggingContext loggingContext)
+        {
+            _expander = _expander.WithLoggingContext(loggingContext);
+        }
```

Comment:
This is unfortunate. Not a big deal and not a blocking comment but it does reduce the cleanliness, IMO. Do instances of this class really use the logging context they are constructed with, before it's replaced here? If the answer is no, can we delay creation of these objects until the final logging context is known?

Reply:
Unfortunately if we batch targets - then buckets are prepared prior the target logging context is created (as that creation logs info about starting run of a target)

Reply:
Gotcha. I took a quick look and I am still wondering if there is an easy way to make it more elegant (and a bit more performant). Currently:

1. `ItemBucket` is created, internally it creates a new `Expander` with the original logging context.
2. (The original logging context is not used, the expander is not used.)
3. `UpdateLoggingContext` is called to replace the expander with a newly created one.

What if, instead, `ItemBucket` construction was split into the real constructor and an `Initialize` method. Only the latter would create the expander with the right logging context. This way there would be no throw-away expander, and the pattern of ctor+Initialize would look slightly more conventional.

Reply:
Looks better then re-creation pattern - thx! - will try that out